File: js-vis/js-impl/odsUtils.js
class OdsUtils {
    static assignNewTags(minNode, nodesCount, minTag, maxTag) {
        let tagOffset = (Math.floor((1 + maxTag - minTag) / nodesCount));
        let tag = minTag;
        let node = minNode;

        while (nodesCount > 0 && node) {
            node.tag = tag;
            node = node.next;
            tag += tagOffset;
            nodesCount -= 1;
        }
    }

    static availableTagAfter(xNode, u) {
        let minTag = xNode.tag + 1;
        let maxTag = u - 1;

        if (xNode.next) {
            maxTag = xNode.next.tag - 1;
        }

        return maxTag - minTag >= 0;
    }

    static getNewTagAfter(xNode, u) {
        let minTag = xNode.tag + 1;
        let maxTag = u - 1;

        if (xNode.next) {
            maxTag = xNode.next.tag - 1;
        }

        let newTag = Math.floor((minTag + maxTag) / 2);
        return newTag;
    }

    static relabel(xNode, n, u, treeIdForHighlight = rootTreeId) {
        let T = this.calculateT(n, u);

        // visualisation.highlight(node => OdsToGraph.isGaphNodeTagInInterval(node.data(), xNode.tag, xNode.tag + 1) && node.data().level === 0 && node.data().treeId === treeIdForHighlight, "blue", true, true);
        visualisation.logMessage("overflow constant T = " + T, "blue", false);
        visualisation.logMessage("searching for smallest enclosing tag interval not in overflow", "blue", false);
        visualisation.addMessageIndent(1);
        let interval = new TagInterval(xNode);
        let minTag1 = interval.minTag;
        let maxTagExcl1 = interval.maxTagExcl;
        visualisation.highlight(node => OdsToGraph.isGaphNodeTagInInterval(node.data(), minTag1, maxTagExcl1, treeIdForHighlight), "blue", false, false);
        interval.increase();

        while (interval.density() > this.overflowThreshold(T, interval.level)) {
            visualisation.logMessage("interval: [" + interval.minTag + "; " + interval.maxTagExcl + "), level = " + interval.level, "blue", true);
            let minTag = interval.minTag;
            let maxTagExcl = interval.maxTagExcl;
            visualisation.highlight(node => OdsToGraph.isGaphNodeTagInInterval(node.data(), minTag, maxTagExcl, treeIdForHighlight), "blue", false, false);
            visualisation.addMessageIndent(1);
            visualisation.logMessage("density = " + interval.density() + " >  overflow threshold = T ^ (-level) = " + this.overflowThreshold(T, interval.level), "blue", false);
            visualisation.addMessageIndent(-1);
            interval.increase();
        }
        visualisation.logMessage("interval: [" + interval.minTag + "; " + interval.maxTagExcl + "), level = " + interval.level, "blue", true);
        visualisation.addMessageIndent(1);
        visualisation.logMessage("density = " + interval.density() + " <=  overflow threshold = T ^ (-level) = " + this.overflowThreshold(T, interval.level), "blue", false)
        visualisation.addMessageIndent(-1);
        let minTag = interval.minTag;
        let maxTagExcl = interval.maxTagExcl;
        visualisation.highlight(node => OdsToGraph.isGaphNodeTagInInterval(node.data(), minTag, maxTagExcl, treeIdForHighlight), "blue", false, false);

        visualisation.addMessageIndent(-1);

        visualisation.logMessage("relabel interval [" + interval.minTag + "; " + interval.maxTagExcl + ")", "blue", true);

        this.assignNewTags(interval.minNode, interval.nodesCount, interval.minTag, interval.maxTagExcl - 1);

        visualisation.refresh(false);
        visualisation.highlight(node => OdsToGraph.isGaphNodeTagInInterval(node.data(), minTag, maxTagExcl, treeIdForHighlight), "blue", true, false);
    }

    static calculateT(n, u) {
        let treeLevel = this.getVirtualTreeLevel(u);
        let T = Math.max(1.001, Math.exp(Math.log(u / n) / treeLevel) - 0.0001);

        return T;
    }

    static overflowThreshold(T, intervalLevel) {
        if (intervalLevel === 0) {
            return 1;
        }

        return Math.pow(T, -intervalLevel);
    }

    static getVirtualTreeLevel(u) {
        return Math.log2(u);
    }

    static violatesInvariant1(n, N) {
        return !(N / 2 <= n && n <= 2 * N);
    }

    static violatesInvariant2(nSublist, NSublist) {
        return !(0 < nSublist && nSublist <= 2 * NSublist);
    }

    static calculateU(N) {
        return Math.max(Math.pow(N, 2), 4 * N);
        // return Math.pow(2, 1 + Math.floor(Math.log2(Math.pow(N, 2))));
    }
}


class TagInterval {
    constructor(xNode) {
        this.level = 0;
        this.minTag = xNode.tag;
        this.maxTagExcl = xNode.tag + 1;
        this.minNode = xNode;
        this.maxNode = xNode;
        this.nodesCount = 1;
    }

    increase() {
        let increaseToTheRight = (Number.parseInt(this.minTag) & Number.parseInt(this.size())) === 0;
        this.level += 1;

        if (increaseToTheRight) {
            this.maxTagExcl = this.minTag + this.size();
            let node = this.maxNode.next;

            while (node !== null && node.tag < this.maxTagExcl) {
                this.maxNode = node;
                node = node.next;
                this.nodesCount += 1;
            }
        } else {
            this.minTag = this.maxTagExcl - this.size();
            let node = this.minNode.prev;

            while (node !== null && node.tag >= this.minTag) {
                this.minNode = node;
                node = node.prev;
                this.nodesCount += 1;
            }
        }
    }

    size() {
        return Math.pow(2, this.level);
    }

    density() {
        return this.nodesCount / this.size();
    }

}

File: python-impl/ods_utils.py
import math

def assign_new_tags(min_node, nodes_to_retag_count, min_tag, max_tag):
    print("assign new tags: ", min_node, nodes_to_retag_count, min_tag, max_tag)
    tag_offset = ((1 + max_tag - min_tag) // nodes_to_retag_count)
    tag = min_tag
    node = min_node
    
    while nodes_to_retag_count > 0 and node is not None:
        node.tag = tag
        
        node = node.next
        tag += tag_offset
        nodes_to_retag_count -= 1

def available_tag_after(x_node, u):
    min_tag = x_node.tag + 1
    if (x_node.next is not None):
        max_tag = x_node.next.tag - 1
    else:
        max_tag = u - 1
    
    return max_tag - min_tag >= 0

def get_new_tag_after(x_node, u):
    min_tag = x_node.tag + 1
    if (x_node.next is not None):
        max_tag = x_node.next.tag - 1
    else:
        max_tag = u - 1
    
    new_tag = (min_tag + max_tag) // 2
    
    return new_tag

def relabel(x_node, n, u):
    print("relabel")
    T = calculate_T(n, u)
    print("T: ", T)
    interval = TagInterval(x_node)
    interval.increase()
    
    while interval.density() > overflow_threshold(T, interval.level):
        print("interval:", interval.min_tag, "-", interval.max_tag_excl, "n:", interval.nodes_count)
        interval.increase()
    
    print("interval:", interval.min_tag, "-", interval.max_tag_excl, "n:", interval.nodes_count)
    assert(interval.max_tag_excl <= u)
    
    assign_new_tags(interval.min_node, interval.nodes_count, interval.min_tag, interval.max_tag_excl - 1)
    
def calculate_u(N):
    return max(int(math.pow(2, math.ceil(math.log2(math.pow(N, 2))))), 4 * N);
    
def get_virtual_tree_level(u):
    return math.log2(u)
    
def violates_invariant_1(n, N):
    return not (N / 2 <= n <= 2 * N)

def violates_invariant_2(n_sublist, N_sublist):
    return not (0 < n_sublist <= 2 * N_sublist)
    
def calculate_T(n, u):
    print("calculate_T with: n:", n, " u:", u)
    tree_level = get_virtual_tree_level(u)
    root_density = (n / u)
    
    T = max(1.001, math.exp(math.log(u/n)/tree_level) - 0.0001)
    
    assert(root_density <= overflow_threshold(T, tree_level))
    
    return T
    
def overflow_threshold(T, interval_level):
    if interval_level == 0:
        return 1
    return math.pow(T, -interval_level)


class TagInterval:
    def __init__(self, x_node):
        self.level = 0
        self.min_tag = x_node.tag
        self.max_tag_excl = x_node.tag + 1
        self.min_node = x_node
        self.max_node = x_node
        self.nodes_count = 1
    
    def increase(self):
        increase_to_the_right = int(self.min_tag) & int(self.size()) == 0
        self.level += 1
        
        if increase_to_the_right:
            self.max_tag_excl = self.min_tag + self.size()
            node = self.max_node.next
            while node is not None and node.tag < self.max_tag_excl:
                self.max_node = node
                node = node.next
                self.nodes_count += 1
        else:
            self.min_tag = self.max_tag_excl - self.size()
            node = self.min_node.prev
            while node is not None and node.tag >= self.min_tag:
                self.min_node = node
                node = node.prev
                self.nodes_count += 1
    
    def size(self):
        return math.pow(2, self.level)
    
    def density(self):
        return self.nodes_count / self.size()

